### ISSUE TITLE
Swap: Stuck after closing approval screen

### ISSUE THREAD

**Issue by VladUXUI** [tallycash/extension#983]
### What browsers are you seeing the problem on?

Chrome

### What were you trying to do?

Approve a new asset for swap

### What did not work?

Clicked approve swap
Approve sign tx popped up
Clicked outside(extension closed)
Came back to extension and went to swaps
That asset now has a "waiting for approval screen"

![CleanShot 2022-02-12 at 12 27 26](https://user-images.githubusercontent.com/80846886/153707638-fdf249a4-f52c-45c1-822e-2c512b2cb3b0.gif)

### Relevant log output

```shell
Version: e2c0d12 (2022-02-11)
```

**Comment by cairomassimo** [comment]
How long did it stay like this? I assume this is the intended behavior: if you go back to the swap page you see the status of the most recent, unfinished swap, which includes waiting for the approval transaction. If this is not the intended behavior, then perhaps we should add the UI to resume/check the currently pending swap. @VladUXUI? Hope this makes sense.

**Comment by cairomassimo** [comment]
Sorry, I misinterpreted the bug. No approval transaction is ever submitted. I managed to reproduce it myself, and it looks bad.

**Comment by VladUXUI** [comment]
It's not be intended behaviour. Because i did not click on final Approve bt, i just clicked outside the extension on the sign-tx page. In normal circumstances, if i would have signed the tx, then it would be a good behaviour.

You can test it out by following my steps above.

Can't tell how long it staid there. I just tested again, and almost 10 minutes later i still see the awaiting for approval

**Comment by VladUXUI** [comment]
Ohh, didn't see your comment :) yes, that is the one

**Comment by cairomassimo** [comment]
I'll be working on this.

**Comment by Shadowfiend** [comment]
Quick note here: a reasonable approach would be to track the transaction hash for a particular contract address when waiting for approval (instead of a boolean), and then using a selector to get that transaction's status to decide whether the transaction is actually waiting for approval (i.e., has no block info and no status), is confirmed (i.e., has a successful status), or is in some other state (dropped/failed).

**Comment by cairomassimo** [comment]
The issue is more profound: _our provider for dApps does not reject transactions if the window is closed, leaving them in an unsettled state forever_. There is no way to restore the sign transaction UI to make progress.
This occurs with external dApps as well, but it's harder to trigger because you have to close the window explicitly. It can happen much more easily with internal transactions, since any click outside the popup will close the window.
![Screenshot from 2022-02-25 10-51-32](https://user-images.githubusercontent.com/4115717/155684888-f069b091-922f-4d53-9f02-4216e0f28fde.png)

**Comment by Shadowfiend** [comment]
> The issue is more profound: our provider for dApps does not reject transactions if the window is closed, leaving them in an unsettled state forever. There is no way to restore the sign transaction UI to make progress.

Talked this through with @VladUXUI a bit [in Flowdock](https://www.flowdock.com/app/cardforcoin/tally/threads/-1AB_twwcd8XIP2XtdyBEvy9fT_). We actually do want to restore the transaction UI in these cases, rather than rejecting transactions on popup close. This will play into transaction queueing as well. Obviously that requires some deeper rework… I would be 100% comfortable if you felt compelled to dive in on that rework. My intention was to make the top-level popup component render `SignTransaction` over the rest of the UI if there was a transaction waiting to be signed, and make sure that transaction approval + rejection both cleared the state. The desired behavior here is that if there's a transaction pending signature, it should be visible in all wallet windows, whether the extension icon popover or any dApp-triggered windows. This will replace our profoundly problematic redirects to the signing route, and will open the door for us to deal with signing resolution better as well.

It should also open the door to manage more state locally instead of in redux for stuff like swap approvals, reducing the complexity of dealing with “remembered” state from previous popup openings.

Tldr, a pending transaction should become a global state that overlays all wallet UI for now.

**Comment by cairomassimo** [comment]
@Shadowfiend it's a good plan. Let's land the fix in #1041 first though, this is quite an annoying bug (especially if you are testing swaps, the only fix is to remove and reinstall the extension). Let me know if you have suggestions about it.
Happy to dive into the rework next :)
